[DOC] User journeys through Tempo (#3716)
grafana/tempo: docs/sources/tempo/introduction/solutions-with-traces/traces-app-insights.md

keywords:
  - Grafana
  - traces
  - tracing
title: Identify bottlenecks and establish SLOs
menuTitle: Identify bottlenecks and establish SLOs
weight: 320
---

# Identify bottlenecks and establish SLOs

Traces give a window into an applications performance, and the insights that go along with that. Using traces, you can discover bottlenecks and potential optimizations that can lead to decreased latency and response times, improving the number of requests the application can handle per second.

## Meet Handy Site Corp

Handy Site Corp, a fake website company, runs an ecommerce application that includes user authentication, a product catalog, order management, payment processing, and other services.

### Define realistic SLOs

Handy Site’s engineers need to establish service level objectives (SLOs) around latency to measure the latency customers experience with the checkout service.
To do this, they can leverage the metrics generated from their span data.

They need to establish realistic targets based on previous history of normal operation modes.
This data helps them identify degradation of service over time.
In addition, they want to be alerted when significant deviations occur.

> ### Utilize span metrics

Reviewer: ```suggestion
### Utilize span metrics to define your SLO and SLI
```

![Metrics generator and exemplars](/media/docs/tempo/intro/traces-metrics-gen-exemplars.png)

Tempo can generate metrics using the [metrics-generator component](https://grafana.com/docs/tempo/latest/metrics-generator/).
These metrics are created based on spans from incoming traces and demonstrate immediate usefulness with respect to application flow and overview.
This includes rate, error, and duration (RED) metrics.

Span metrics can provide in-depth monitoring of your system. The generated metrics show application-level insight into your monitoring, as traces propagate through your application's services.

Span metrics lower the entry barrier for using exemplars.
An [exemplar](https://grafana.com/docs/grafana/latest/basics/exemplars/) serves as a detailed example of one of the observations aggregated into a metric. An exemplar contains the observed value together with an optional timestamp and arbitrary trace IDs, which are typically used to reference a trace.
Since traces and metrics co-exist in the metrics-generator, exemplars can be automatically added, providing additional value to these metrics.

> ### Monitor latency

Reviewer: ```suggestion
```
Not sure this is needed as a separate header?

Author: Let's see how the page looks without the heading. Diving the content into sections help readability. Open to suggestions.

> In this case, Handy Site wants to monitor latency.

Reviewer: ```suggestion
Handy Site decides they're most interested in monitoring the latency of requests processed by their checkout service and want to set an objective that 99.5% of requests in a given month should complete within 2 seconds.
```

> They can leverage the `traces_spanmetrics_latency` metric with the corresponding labels, such as `service name = checkoutservice`.

Reviewer: ```suggestion
To define a service level indicator (SLI) that they can use to track their progress against their objective they use the `traces_spanmetrics_latency` metric with the proper label selectors, such as `service name = checkoutservice`.
```

> For example, if there are specific spans of interest such as release versions, per-service endpoints, and others, you can create and use those metrics as extra dimensions for correlations.

Reviewer: ```suggestion
The metrics-generator adds a default set of labels to the metrics it generates, including `span_kind` and `status_code`. However, if they were interested in calculating checkout service latency per endpoint or per version of the software, they could change the configuration of the Tempo metrics-generator to add these custom dimensions as labels to their spanmetrics. 
```

Author: Your suggestions are fantastic. Thank you.

> With all of this in place, Handy Site uses the data generated with metrics-generator in [Grafana SLO](https://grafana.com/docs/grafana-cloud/alerting-and-irm/slo/) to establish an [SLI](https://grafana.com/docs/grafana-cloud/alerting-and-irm/slo/create/).

Reviewer: ```suggestion
With all of this in place, Handy Site now opens the [Grafana SLO](https://grafana.com/docs/grafana-cloud/alerting-and-irm/slo/) application and follows the setup flow to establish an [SLI](https://grafana.com/docs/grafana-cloud/alerting-and-irm/slo/create/) for their checkout service around the `traces_spanmetrics_latency` metric..
```

> They can now be alerted to degradations in service quality that directly impacts their end user experience.

Reviewer: ```suggestion
They can now be alerted to degradations in service quality that directly impact their end user experience. SLO-based alerting also ensures that they don't suffer from noisy alerts. Alerts are only triggered when the value of the SLI is such that the team is in danger of missing their SLO.
```
